brave/src/main/java/brave/handler/SpanListener.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.handler;
+
+import brave.Span;
+import brave.internal.Nullable;
+import brave.propagation.TraceContext;
+import java.lang.ref.WeakReference;
+
+/**
+ * This is like {@link FinishedSpanHandler}, except it can cover all state conditions, including
+ * when a span is created or abandoned. The purpose of this type is to allow tracking of children,
+ * or partitioning of data for backend that needs to see an entire {@linkplain
+ * TraceContext#localRootId() local root}.
+ *
+ * <p>As with {@link FinishedSpanHandler}, it is important to do work quickly as callbacks are run
+ * on the same thread as application code. That said, there are some rules to keep in mind below.
+ *
+ * <p>The {@link TraceContext} parameter from {@link #onCreate} will be the same reference for
+ * all callbacks, except {@link #onOrphan}, which has value, but not reference equality.
+ *
+ * <p>The {@link MutableSpan} parameter from {@link #onCreate} will be the same reference for
+ * all callbacks. Do not mutate {@link MutableSpan} between callbacks as it is not thread safe.
+ *
+ * <p>If caching the {@link TraceContext} parameter, consider a {@link WeakReference} to avoid
+ * holding up garbage collection.
+ */
+public class SpanListener {
+  /** Use to avoid comparing against null references */
+  public static final SpanListener NOOP = new SpanListener() {
+    @Override public String toString() {
+      return "NoopSpanListener{}";
+    }
+  };
+
+  protected SpanListener() {
+  }
+
+  /**
+   * This is called when a span is allocated, but before it is started. An allocation here will
+   * result in one of:
```

Comment:
One of the following will be called later when the created span is no longer in use.

Reply:
Currently I was surprised since it reads like these are called right away